examples/petsc/bps.c

```diff
@@ -385,26 +385,40 @@ static PetscErrorCode Run(RunParams rp, PetscInt num_resources,
 
   PetscFunctionBeginUser;
   // Setup DM
-  if (rp->read_mesh) {
-    ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, rp->filename, NULL, PETSC_TRUE,
-                                &dm);
-    CHKERRQ(ierr);
-  } else {
-    if (rp->user_l_nodes) {
-      // Find a nicely composite number of elements no less than global nodes
-      PetscMPIInt size;
-      ierr = MPI_Comm_size(rp->comm, &size); CHKERRQ(ierr);
-      for (PetscInt g_elem =
-             PetscMax(1, size * rp->local_nodes / PetscPowInt(rp->degree, rp->dim));
-           ;
-           g_elem++) {
-        Split3(g_elem, rp->mesh_elem, true);
-        if (Max3(rp->mesh_elem) / Min3(rp->mesh_elem) <= 2) break;
+  if (0) {
+    if (rp->read_mesh) {
+      ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, rp->filename, NULL, PETSC_TRUE,
+                                  &dm);
+      CHKERRQ(ierr);
+    } else {
+      if (rp->user_l_nodes) {
+        // Find a nicely composite number of elements no less than global nodes
+        PetscMPIInt size;
+        ierr = MPI_Comm_size(rp->comm, &size); CHKERRQ(ierr);
+        for (PetscInt g_elem =
+               PetscMax(1, size * rp->local_nodes / PetscPowInt(rp->degree, rp->dim));
+             ;
+             g_elem++) {
+          Split3(g_elem, rp->mesh_elem, true);
+          if (Max3(rp->mesh_elem) / Min3(rp->mesh_elem) <= 2) break;
+        }
       }
+      ierr = DMPlexCreateBoxMesh(PETSC_COMM_WORLD, rp->dim, PETSC_FALSE,
+                                 rp->mesh_elem,
+                                 NULL, NULL, NULL, PETSC_TRUE, &dm); CHKERRQ(ierr);
+      ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
```

Comment:
Here. If `SetFromOptions` get called here, does that help?

Reply:
No, I get error. I think the problem is in `DMPlexCreateBoxMesh` since I copied the block of the code from ratel to create `fe_coord` even didn't change the name of variables.

Reply:
My general approach here would be to look at the source for both and start looking for differences. I'm working on my PP slides, but I can take a look tomorrow.

Reply:
@rezgarshakeri `DMPlexBoxMesh()` does not process options. It only uses the inputs and is for people who want to use purely APIs. `DMSetFromOptions()` is the way to use the options, and I think should be the preferred way to create meshes.

Reply:
Right, that's what we do in Ratel. The hard thing is that in this example we want to loop over a range of target local vector sizes and create DMs for box meshes with different numbers of faces inside of that loop. Doing that with DMSetFromOptions is awkward, as far as I can tell.

Reply:
We have an arbitrary number at runtime, an the logic for generating the sizes is inside the C code.